Source/Scene/ShadowMapShader.js

```diff
@@ -15,79 +15,175 @@ define([
+            var define = defines[i];
+            if (define.indexOf('ENABLE_VERTEX_LIGHTING') !== -1 || define.indexOf('ENABLE_DAYNIGHT_SHADING') !== -1) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function terrainHasNormalVarying(defines) {
+        var length = defines.length;
+        for (var i = 0; i < length; ++i) {
+            var define = defines[i];
+            if (define.indexOf('ENABLE_VERTEX_LIGHTING') !== -1) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    ShadowMapShader.createShadowCastVertexShader = function(vs, isPointLight, isTerrain) {
+        var defines = vs.defines.slice(0);
+        var sources = vs.sources.slice(0);
+
+        if (isTerrain && !terrainHasPositionVarying(defines)) {
+            defines.push('ENABLE_DAYNIGHT_SHADING');
+        }
+
+        var positionVaryingName = findPositionVarying(vs);
+        var hasPositionVarying = defined(positionVaryingName);
 
         if (isPointLight && !hasPositionVarying) {
-            vs = ShaderSource.replaceMain(vs, 'czm_shadow_main');
-            vs +=
+            var length = sources.length;
+            for (var j = 0; j < length; ++j) {
+                sources[j] = ShaderSource.replaceMain(sources[j], 'czm_shadow_cast_main');
+            }
+
+            var shadowVS =
                 'varying vec3 v_positionEC; \n' +
                 'void main() \n' +
                 '{ \n' +
-                '    czm_shadow_main(); \n' +
+                '    czm_shadow_cast_main(); \n' +
                 '    v_positionEC = (czm_inverseProjection * gl_Position).xyz; \n' +
                 '}';
+            sources.push(shadowVS);
         }
 
-        return vs;
+        return new ShaderSource({
+            defines : defines,
+            sources : sources
+        });
     };
 
-    ShadowMapShader.createShadowCastFragmentShader = function(fs, frameState, opaque, positionVaryingName) {
-        // TODO : is there an easy way to tell if a model or primitive is opaque before going here?
-        var hasPositionVarying = defined(positionVaryingName) && (fs.indexOf(positionVaryingName) > -1);
-        positionVaryingName = hasPositionVarying ? positionVaryingName : 'v_positionEC';
-        var isPointLight = frameState.shadowMap._isPointLight;
-        var usesDepthTexture = frameState.shadowMap._usesDepthTexture;
+    ShadowMapShader.createShadowCastFragmentShader = function(fs, isPointLight, useDepthTexture, opaque, isTerrain) {
+        var defines = fs.defines.slice(0);
+        var sources = fs.sources.slice(0);
+
+        var positionVaryingName = findPositionVarying(fs);
+        var hasPositionVarying = defined(positionVaryingName);
+        if (!hasPositionVarying) {
+            positionVaryingName = 'v_positionEC';
+        }
+
+        if (isTerrain && !terrainHasPositionVarying(defines)) {
+            defines.push('ENABLE_DAYNIGHT_SHADING');
+        }
```

Comment:
You don't need to add the define to the fragment shader. (Similar comment as before)